### PR TITLE
make teslas great again

### PR DESCRIPTION
**Pull request by MilonPL**
## About the PR

Added a passive drain to teslas (3 per second, might need to be adjusted). Once it reaches 100 power, a bunch of lightning balls spawn and power goes back to 0. If it reaches -100, the tesla despawns.
Also replaced the ChasingWalk with RandomWalk so it has randomized pathfinding like the singulo.
Uncommented the spawns, cargo can order it too.

## Why / Balance

There was a bunch of discussion about it before, I think it ended up with "wyci" so I coded it. Yippee. 
The main point against having teslas was that it was a guaranteed round end, so well, it's not a guaranteed round end anymore. Also you need the PA to be constantly on, like with the singulo, or it's just going to die. Can't wait for engi to fuck up.

## Technical details

Uncomments spawns, minor changes to the yaml file, ~~I tried to avoid touching upstream code so this only adds two datafields to the component (which could be done with a separate component too but that feels like just bloat at this point), the actual updates are being handled by a separate system using the public AdjustEnergy() method. Had to add access to the component to modify the DataFields, again, could be all done in a separate TeslaPassiveEnergyDrainComponent instead or whatever.~~
Decided to change the system to partial instead and add TeslaEnergyBallSystem.PassiveDrain.cs on in the DeltaV namespace to avoid having to call it and also avoid the access issues. Maybe I overcomplicated it. I dunno.

## Requirements

- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes

**Changelog**

:cl:
- add: Re-added teslas!
- tweak: Tesla now has a random pathfinding and a passive energy drain. Don't forget to leave the PA on!

### CODE REVIEW

**Comment by Avalon-Proto** [comment]
THE GRAND RETURN OF LORD TESLOTH!

**Comment by dotCatshark** [comment]
consider changing it so that the tesla only loses energy when it eats something. i think making it require the pa to run removes some of the uniqueness of it. you might be able to cannibalize the code for that from the singulo code (singulo should gain energy as it consumes stuff. if it  doesnt then that particular mechanic from 13 wasnt ported... for some reason)

**Comment by MilonPL** [comment]
> consider changing it so that the tesla only loses energy when it eats something. i think making it require the pa to run removes some of the uniqueness of it. you might be able to cannibalize the code for that from the singulo code (singulo should gain energy as it consumes stuff. if it doesnt then that particular mechanic from 13 wasnt ported... for some reason)

All you'd have to do is flip `ConsumeStuffEnergy` from 2 to -2 technically. But then the tesla is kinda guaranteed to fade away pretty soon.

**Comment by deltanedas** [comment]
if it lost energy by consuming i think it would be fine to keep the targeted walk since it would go after power infrastructure etc then die

**Comment by MilonPL** [comment]
@deltanedas this is ready for review fyi, game directors fake anyways and I haven't heard any other ideas regarding how to implement this

File: Content.Server/DeltaV/Tesla/TeslaEnergyBallSystem.PassiveDrain.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        var curTime = _timing.CurTime;
+        var query = EntityQueryEnumerator<TeslaEnergyBallComponent>();
+
+        while (query.MoveNext(out var uid, out var component))
+        {
+            if (curTime < component.NextUpdateTime)
+                continue;
+
+            component.NextUpdateTime = curTime + TimeSpan.FromSeconds(1);
```
Comment (deltanedas) [inline Content.Server/DeltaV/Tesla/TeslaEnergyBallSystem.PassiveDrain.cs]: have the timespan be a system constant or a component field

File: Content.Server/Tesla/Components/TeslaEnergyBallComponent.cs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 using Content.Server.Tesla.EntitySystems;
 using Robust.Shared.Audio;
 using Robust.Shared.Prototypes;
+using Robust.Shared.Timing; // DeltaV
```
Comment (deltanedas) [inline Content.Server/Tesla/Components/TeslaEnergyBallComponent.cs:5]: put it in the other file

File: Content.Server/Tesla/Components/TeslaEnergyBallComponent.cs [hunk 3]
```diff
@@ -51,4 +52,18 @@ public sealed partial class TeslaEnergyBallComponent : Component
...
+    /// The amount of energy drained passively per update.
+    /// </summary>
+    [DataField]
+    public float PassiveEnergyDrainRate = 3f;
+
+    /// <summary>
+    /// The <see cref="IGameTiming.CurTime"/> timespan of next update.
+    /// </summary>
+    [DataField]
+    public TimeSpan NextUpdateTime = default!;
```
Comment (deltanedas) [inline Content.Server/Tesla/Components/TeslaEnergyBallComponent.cs]: TimeSpan.Zero

File: Content.Server/Tesla/Components/TeslaEnergyBallComponent.cs [hunk 4]
```diff
@@ -51,4 +52,18 @@ public sealed partial class TeslaEnergyBallComponent : Component
...
+    /// <summary>
+    /// The amount of energy drained passively per update.
+    /// </summary>
+    [DataField]
+    public float PassiveEnergyDrainRate = 3f;
+
+    /// <summary>
+    /// The <see cref="IGameTiming.CurTime"/> timespan of next update.
+    /// </summary>
+    [DataField]
```
Comment (deltanedas) [inline Content.Server/Tesla/Components/TeslaEnergyBallComponent.cs]: use TimeOffsetSerializer and use AutoPausedField+AutoGenerateComponentPause

**Review by deltanedas — CHANGES_REQUESTED** [review]
i forgot to request changes ages ago

**Comment by MilonPL** [comment]
> i forgot to request changes ages ago

no fucking way i though you just hated it

**Comment by deltanedas** [comment]
forgot to click a button highly real

File: Content.Server/Tesla/Components/TeslaEnergyBallComponent.cs (hunk 2 above)
Comment (MilonPL) [reply Content.Server/Tesla/Components/TeslaEnergyBallComponent.cs:5]: ??? 
needed for `IGameTiming.CurTime` in the summary

Comment (deltanedas) [reply Content.Server/Tesla/Components/TeslaEnergyBallComponent.cs:5]: real

**Review by deltanedas — APPROVED** [review]

**Pull request merged**